File: ruby/Damage.rb
require_relative 'WeaponType'
require_relative 'Weapon'
require_relative 'DamageToUI'

module Deepspace
    class Damage
        def initialize(s)
            @nShields = s
        end

        # Devuelve el nº de escudos ajustado a las colección de
        # potenciadores de escudos suministrada como parámetro.
        # @param Array<ShieldBooster> s
        # @return Int nShieldsNew
        def adjust (s)
            [s.length, @nShields].min
        end

        # Decrementa en una unidad el número de potenciadores de escudo que
        # deben ser eliminados. Ese contador no puede ser inferior a cero
        # en ningún caso
        def discardShieldBooster
            if @nShields > 0
                @nShields -= 1
            end
        end

        # Devuelve true si el daño representado no tiene ningún efecto. Esto
        # quiere decir que no implica la pérdida de ningún tipo de accesorio
        # (armas o potenciadores de escudo)
        # @return boolean
        def hasNoEffect
            @nShields==0
        end


        # Métodos get
        attr_reader :nShields


        def getUIversion
            DamageToUI.new(self)
        end
        
        # Representación del objeto en String
        def to_s
            "nShields: #{@nShields}"
        end

        private_class_method :new
    end
end

File: java/DeepSpace/src/deepspace/Damage.java
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

abstract class Damage {
    private int nShields;
    
    Damage(int s){
        nShields = s;
    }
    
    // Método de copia
    public abstract Damage copy();
    
    // Devuelve una versión ajustada del objeto a las colecciones de armas y
    // potenciadores de escudos suministradas como parámetro.
    // Partiendo del daño representado por el objeto que recibe este mensaje,
    // se devuelve una copia del mismo pero reducida si es necesario para que
    // no implique perder armas o potenciadores de escudos que no están en
    // las colecciones de los parámetros
    // @param Array<Weapon> w
    // @param Array<ShieldBooster> s
    // @return Damage damage
    public abstract Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s);
    
    
    // Elimina el arma pasada como parámetro
    // @param Weapon w
    public abstract void discardWeapon(Weapon w);
    
    // Decrementa en una unidad el número de potenciadores de escudo que
    // deben ser eliminados. Ese contador no puede ser inferior a cero
    // en ningún caso
    public void discardShieldBooster(){
        if (nShields > 0)
            nShields--;
    }
    
    // Devuelve true si el daño representado no tiene ningún efecto. Esto
    // quiere decir que no implica la pérdida de ningún tipo de accesorio
    // (armas o potenciadores de escudo)
    // @return boolean
    public abstract boolean hasNoEffect();
    
    
    public int getNShields(){
        return nShields;
    }
    
    abstract DamageToUI getUIversion();
    
    // Representación del objeto en String
    public abstract String toString();
}